### ISSUE TITLE
dnsssec brakes sooner or later - centos 9 stream

### ISSUE THREAD

**Issue by lejeczek** [freeipa/freeipa-container#538]
Hi guys.

I've reproduced this over dozen of times - each time, when both container & IPA deployed successfully, after some time DNSSEC brakes with:
...
31-May-2023 05:32:21.626 info: client @0x7f9938b0b068 127.0.0.1#60304 (google.com): query failed (broken trust chain) for wp.pl/IN/A at ../../../lib/ns/query.c:7355
...
and that is every domain & every query but for its own domain.
Sometimes DNSSEC brakes as easily as within minute, very first 'ipactl restart' was run, or ipa backup was done or just container was restarted.
Sometimes it takes longer or more than just first "restart" but soon enough it happens.

Both, in my case, host & container are Centos 9 Stream. I've tried different 'podman' versions. Network is 'macvlan'. An example of a 'run':

-> podman run -dt --security-opt label=disable --privileged --dns=none --add-host=swir-ipa.dupa.kupa:10.4.1.200 --tz=local --network off-host --ip 10.4.1.200 --volume /devs/ROOTFUL.containers/freeipa.dupa.kupa/root:/root:Z --volume /devs/ROOTFUL.containers/freeipa.dupa.kupa/data:/data:Z --hostname swir-ipa.dupa.kupa --name freeipa.dupa.kupa localhost/freeipa-server ipa-server-install --setup-dns --allow-zone-overlap --setup-kra --forwarder=208.67.222.222 --forwarder=208.67.220.220 --forwarder=8.8.8.8 --ds-password=7#dirsrv --admin-password=7#diradm --ip-address=10.4.1.200 --domain=dupa.kupa --realm=DUPA.KUPA --unattended

IPA's DNS after deployment - forwarders or no forwarders, makes no difference.
Networking is fully working and time sync is not issue.

Destroy such broken IPA container & immediately re-run with the same exact 'args' in the same very place - and DNSSEC works a okey only until...

many thanks, L.

**Comment by adelton** [comment]
Similar issues were reported before (https://pagure.io/freeipa/issue/8864, https://bugzilla.redhat.com/show_bug.cgi?id=1999321) making me think this is a generic DNS server / bind / FreeIPA issue, not really related to the FreeIPA containerization.

**Comment by adelton** [comment]
On a possibly unrelated note: where did you see the `--security-opt label=disable --privileged` options recommended for running the FreeIPA server container?

**Comment by lejeczek** [comment]
Well. I'd like to agree yet that "same" IPA does not "suffer" from this issues when on bare-metal or KVM - inasmuch as much ostree/packages are the same or not.
`--privileged` just for sake of an argument I added in latest 'run' iteration.
`--security-opt label=disable` gives access to mapped vol data copied of host - otherwise such data is inaccessible in container, as I understand/see it.

Bit more feedback, just got first master but - fedora-37 seems (on this same everything, centos 9, etc.) to work with (those) issues, is a okey.
I'm adding more replicas and will get back in with more feedback.

**Comment by adelton** [comment]
Using `--privileged` is actively harmful.

As for the access to the volume data -- the `:Z` takes care of the SELinux context, so that `label=disable` should never be needed.

**Comment by lejeczek** [comment]
Yes, seems that f37 is free from this issue.
What I'm doing - I'm thinking & thinking & I reckon - is nothing unusual, vanilla-default, should be easily reproducible.

If it's not container - which working f37 container might, not confirm but suggest - then how would that play into the fact that on metal/kvm Centos 9 IPA does not show this very problem.

**Comment by adelton** [comment]
@abbra Would you have any insight into that DNSSEC issue?

**Comment by abbra** [comment]
No. I don't think it is DNSSEC issue. It might be a problem with restoring actual bind configuration and thus defaulting to DNSSEC validation. E.g. `/etc/named/ipa-ext.conf` where we set `dnssec-validation no;` might be missing or unreadable after restoring the backup.

**Comment by adelton** [comment]
Ah, I read it as @lejeczek makes a backup, not restores it ...

**Comment by abbra** [comment]
If that happens only after backup, the same issue stays -- something is damaging a configuration. At this point @lejeczek claims it only happens in the containerized environment, not in VMs, so presumably it is an environmental problem in what backup/restore scripts expect and what containers do provide. Perhaps selinux labeling is wrong or some of symlink magic is broken as we do not expect to have those symlinks in backup process.

**Comment by adelton** [comment]
Got it. So @lejeczek, we will likely need an exact reproducer -- what minimal steps _exactly_ do you do, from starting the fresh new container to the error message?

**Comment by lejeczek** [comment]
Hi. Pretty default I think, as I show above, but... also with no 'forwarders' - though such one "misbehaves" differently

HOST:
-> $ podman images
REPOSITORY                TAG         IMAGE ID      CREATED      SIZE
localhost/freeipa-server  latest      8d1c5aa2173d  2 weeks ago  1.07 GB
quay.io/centos/centos     stream9     94fe66853595  2 weeks ago  160 MB

-> $ { export _DOM=dupa.kupa _H=swir-ipa _IP=10.3.1.228; _NAME="freeipa.${_DOM}"; export _PATH=${HOME}/containers/FLATfiles/${_NAME}; echo; mkdir -p ${_PATH}/{,root,data}; }; podman run -dt --security-opt label=disable --dns=none --add-host=${_H}.${_DOM}:${_IP} --tz=local --network off-host --ip ${_IP} --volume ${_PATH}/root:/root:Z --volume ${_PATH}/data:/data:Z --hostname ${_H}.${_DOM} --name ${_NAME} localhost/freeipa-server ipa-server-install --setup-dns --no-forwarders --allow-zone-overlap --setup-kra --ds-password=7#dirsrv --admin-password=7#diradm --ip-address=${_IP} --domain=${_DOM} --realm=${_DOM^^} --unattended
...
[  OK  ] Finished Configure IPA server upon the first start.

CONTAINER is c9s, after IPA completed deployment successfully:
-> $ host wp.pl
wp.pl has address 212.77.98.9
wp.pl mail is handled by 5 mx5.wp.pl.
wp.pl mail is handled by 0 mx.wp.pl.
-> $ ipa-backup 
Preparing backup on swir-ipa.dupa.kupa
Local roles match globally used roles, proceeding.
Stopping IPA services
Backing up ipaca in DUPA-KUPA to LDIF
Backing up userRoot in DUPA-KUPA to LDIF
Backing up DUPA-KUPA
Backing up files
Starting IPA service
Backed up to /var/lib/ipa/backup/ipa-full-2023-06-10-13-59-05
The ipa-backup command was successful
-> $ host wp.pl
Host wp.pl not found: 2(SERVFAIL)

-> $ tailf /var/named/data/query_errors.log 
...
10-Jun-2023 13:59:21.560 info: client @0x7f818caa64a8 127.0.0.1#45783 (wp.pl): query failed (broken trust chain) for wp.pl/IN/A at ../../../lib/ns/query.c:7355

-> $ ping 212.77.98.9
PING 212.77.98.9 (212.77.98.9) 56(84) bytes of data.
64 bytes from 212.77.98.9: icmp_seq=1 ttl=59 time=19.4 ms
64 bytes from 212.77.98.9: icmp_seq=2 ttl=59 time=17.4 ms

**Comment by adelton** [comment]
Please apply temporarily the following patch
```diff
diff --git a/tests/run-master-and-replica.sh b/tests/run-master-and-replica.sh
index 282eef5..03fceb0 100755
--- a/tests/run-master-and-replica.sh
+++ b/tests/run-master-and-replica.sh
@@ -135,6 +135,7 @@ if [ "$(id -u)" != 0 -a "$docker" == podman -a "$replica" != none ] ; then
 elif [ "$readonly" == "--read-only" ] ; then
        readonly_run="$readonly --dns=127.0.0.1"
 fi
+readonly_run="--dns=none"
 
 fresh_install=true
 if [ -f "$VOLUME/build-id" ] ; then
```
and test with `tests/run-master-and-replica.sh`.

Specifically, this is what I see:

```
$ docker=podman replica=none tests/run-master-and-replica.sh quay.io/freeipa/freeipa-server:centos-9-stream
[...]
Restarting the directory server
The ipa-kra-install command was successful
OK tests/run-master-and-replica.sh.

$ podman exec freeipa-master host wp.pl
wp.pl has address 212.77.98.9
wp.pl mail is handled by 0 mx.wp.pl.
wp.pl mail is handled by 5 mx5.wp.pl.

$ podman exec freeipa-master ipa-backup
Preparing backup on ipa.example.test
Local roles match globally used roles, proceeding.
Stopping IPA services
Backing up ipaca in EXAMPLE-TEST to LDIF
Backing up userRoot in EXAMPLE-TEST to LDIF
Backing up EXAMPLE-TEST
Backing up files
Starting IPA service
Backed up to /var/lib/ipa/backup/ipa-full-2023-06-10-13-43-06
The ipa-backup command was successful

$ podman exec freeipa-master host wp.pl
wp.pl has address 212.77.98.9
wp.pl mail is handled by 0 mx.wp.pl.
wp.pl mail is handled by 5 mx5.wp.pl.

$ podman exec freeipa-master cat /etc/resolv.conf 
# auto-generated by IPA installer
search example.test
nameserver 127.0.0.1
```

If these steps pass for you just like they pass for me, start tweaking the `tests/run-master-and-replica.sh` (run `podman rm -f freeipa-master` before each retry) to get the setup closer to what you are using, to find out where things start to break.
